internals/overlord/logstate/loki.go

```diff
@@ -0,0 +1,157 @@
+// Copyright (c) 2023 Canonical Ltd
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License version 3 as
+// published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package logstate
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"sort"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/canonical/pebble/internals/plan"
+	"github.com/canonical/pebble/internals/servicelog"
+)
+
+const (
+	maxLokiRequestSize = 10
+	lokiRequestTimeout = 10 * time.Second
+)
+
+type lokiClient struct {
+	target    *plan.LogTarget
+	remoteURL string
+	// buffered entries are "sharded" by service name
+	entries    map[string][]lokiEntry
+	numEntries int
+
+	httpClient http.Client
+}
+
+func newLokiClient(target *plan.LogTarget) logClient {
+	return &lokiClient{
+		target:     target,
+		remoteURL:  target.Location,
+		entries:    map[string][]lokiEntry{},
+		httpClient: http.Client{Timeout: lokiRequestTimeout},
+	}
+}
+
+func (c *lokiClient) Write(ctx context.Context, entry servicelog.Entry) error {
+	c.entries[entry.Service] = append(c.entries[entry.Service], asLokiEntry(entry))
+	c.numEntries++
+	if c.numEntries >= maxLokiRequestSize {
+		return c.Flush(ctx)
+	}
+	return nil
+}
+
+func asLokiEntry(entry servicelog.Entry) lokiEntry {
+	return lokiEntry{
+		strconv.FormatInt(entry.Time.UnixNano(), 10),
+		strings.TrimSuffix(entry.Message, "\n"),
+	}
+}
+
+func (c *lokiClient) Flush(ctx context.Context) error {
+	if c.numEntries == 0 {
+		return nil // no-op
+	}
+	defer c.emptyBuffer()
+
+	req := c.buildRequest()
+	jsonReq, err := json.Marshal(req)
+	if err != nil {
+		return fmt.Errorf("encoding request to JSON: %v", err)
+	}
+
+	httpReq, err := http.NewRequestWithContext(ctx, http.MethodPost, c.remoteURL, bytes.NewReader(jsonReq))
+	if err != nil {
+		return fmt.Errorf("creating HTTP request: %v", err)
+	}
+	httpReq.Header.Set("Content-Type", "application/json")
+
+	resp, err := c.httpClient.Do(httpReq)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		// Drain request body to allow connection reuse
+		// see https://pkg.go.dev/net/http#Response.Body
+		_, _ = io.Copy(io.Discard, io.LimitReader(resp.Body, 1024*1024))
+		_ = resp.Body.Close()
+	}()
+
+	// Check response status code to see if it was successful
+	if code := resp.StatusCode; code < 200 || code >= 300 {
+		// Request to Loki failed
+		b, err := io.ReadAll(io.LimitReader(resp.Body, 1024))
+		if err != nil {
+			b = append(b, []byte("//couldn't read response body: ")...)
```

Comment:
Why does this message start with ```//``` ?

Reply:
To indicate it's not the response body

Reply:
OK, sorry, let me ask differently. Would it not be cleaner to add two newlines before the final error message, so that its on its own line? Or is this a convention?

Reply:
No, this is not a convention. The point is to have a different error message if we can't read the response body.
```
cannot send logs to Loki (HTTP 503), response body:
//couldn't read response body: i/o error
```
vs
```
cannot send logs to Loki (HTTP 503), response body:
The server is currently unavailable.
```

Reply:
Sorry for going on with this, this is my last comment, I know this is a trivial point.

The thing that bugs me is that the two errors are completely unrelated (and that under normal circumstances, we discard errors related to the body read failing under the ```defer func()``` path), but yet we emit an error that contains both errors in the same message (for example ```HTTP 503 .... i/o error```). Also the response may only fail halfway through the read, so that you see a partial body, and then an error saying that the body cannot be read. I know in practise the body is often super useful for debugging, so my only question is whether the body (for ```len(b) > 0```) should not instead be sent to ```logger.Debugf``` or ```logger.Noticef```.

Reply:
What's your proposed solution here then? :)

If the request returns a non-2xx status code, then semantically that's an error response, so we want to return an error from this function. Just the status code on its own may not be descriptive enough (e.g. 400 or 500), so we'd better read the response body, in order to provide some more useful information to the user.

However, **the act of reading the response itself** can fail and return an error. What should we do with this error? What I wrote seems like a reasonable approach to me - although I'm certainly open to discussing alternatives.

As a rule, I'm trying not to log errors inside the client - we should instead pass them up to the gatherer, so the gatherer can act on them (and log them) if required.

Reply:
Yeah, given the above, I think what @barrettj12 has is fine here. We could tweak the syntax, but `//` seems like a reasonable colour for the bikeshed.